### ISSUE TITLE
Refactor and unify JabRef OO components

### ISSUE THREAD

**Issue by subhramit** [JabRef/jabref#11829]
## Project: Refactor JabRef-OpenOffice components
There is a lot of scope for improvement in code quality in JabRef's OpenOffice/LibreOffice integration components to make it much more maintainable. These include, but are not limited to:
- [ ] Unification of error handling mechanism for JabRef-OOdoc interaction: Currently there exists a mix of returning Optional, throwing exceptions, and using `OOResult`/`OOVoidResult`. We can standardize the error handling approaches  for consistency.
- [ ] Use of Dependency Injection
- [ ] Break down large methods like `guiActionInsertEntry()`
- [ ] Use more constants instead of magic strings e.g. as used in the error messages
- [ ] Use of built-in functional interfaces: for methods that take `Supplier<OOError>` as parameters - will improve code clarity
- [ ] Use of modern Java features like streams and optional chaining
- [ ] Separation of concerns: classes such as `OOBibBase` handle both business logic and UI interactions. We can consider extracting UI-related code into a separate class or layer.
- [ ] Inner classes like `ComparableMark` can be converted to record classes.
- [ ] Unification of backend for CSL and JStyles (Optional, and a big project - I am fine with all the CSL logic being separate due to the difference in code philosophy and time required for unification).

I will be working on this, so will be using this issue to track the state/progress.
The list is not yet complete, so will be updated from time to time as I come across more scopes for improvements.

**Comment by leaf-soba** [comment]
I want take the task "Inner classes like ComparableMark can be converted to record classes."

**Comment by anuv-bit** [comment]
Hi @subhramit , we're a group of students working on a university software engineering course project and would love to tackle this issue. Could you assign this to us?

**Comment by subhramit** [comment]
Hey, just to be transparent, please go through our contributing guidelines and policies on AI. If the quality of PRs don't match expected standards, we may close pull request(s).

**Comment by sam-r914** [comment]
Hi @subhramit, I'm one of the members of students working on this task with anuv-bit. Would you be able to elaborate some more on what you are referring to in the unification of error handling subtask by 'standardizing the error handling approaches'? Are you looking to minimize the use of one or more of the error handling methods in the Open Office files? Or are you wanting to have definitive approaches for certain task contexts, like using Optional for retrieving document/citation information?

**Comment by subhramit** [comment]
> Hi [@subhramit](https://github.com/subhramit), I'm one of the members of students working on this task with anuv-bit. Would you be able to elaborate some more on what you are referring to in the unification of error handling subtask by 'standardizing the error handling approaches'? Are you looking to minimize the use of one or more of the error handling methods in the Open Office files? Or are you wanting to have definitive approaches for certain task contexts, like using Optional for retrieving document/citation information?

That sub-task would be a bit difficult unless you have the context of the whole LO-doc-JabRef interaction. Try to investigate what `OOVoid` and `OOResult` are used in error handling - and compare them to simply throwing exceptions.
Check if these can be re-used at other places where there is interaction with the document. For example, in methods where bibliography is generated.

**Comment by sam-r914** [comment]
> > Hi [@subhramit](https://github.com/subhramit), I'm one of the members of students working on this task with anuv-bit. Would you be able to elaborate some more on what you are referring to in the unification of error handling subtask by 'standardizing the error handling approaches'? Are you looking to minimize the use of one or more of the error handling methods in the Open Office files? Or are you wanting to have definitive approaches for certain task contexts, like using Optional for retrieving document/citation information?
> 
> That sub-task would be a bit difficult unless you have the context of the whole LO-doc-JabRef interaction. Try to investigate what `OOVoid` and `OOResult` are used in error handling - and compare them to simply throwing exceptions. Check if these can be re-used at other places where there is interaction with the document. For example, in methods where bibliography is generated.

Thanks for your response! From my understanding, you are wanting for throwing exceptions to be replaced (where it makes sense and doesn't screw up the workflow) with OOVoid and OOResult error handling methods so that there is less deviation in error handling methods throughout the Open Office components?

**Comment by subhramit** [comment]
> > > Hi [@subhramit](https://github.com/subhramit), I'm one of the members of students working on this task with anuv-bit. Would you be able to elaborate some more on what you are referring to in the unification of error handling subtask by 'standardizing the error handling approaches'? Are you looking to minimize the use of one or more of the error handling methods in the Open Office files? Or are you wanting to have definitive approaches for certain task contexts, like using Optional for retrieving document/citation information?
> > 
> > 
> > That sub-task would be a bit difficult unless you have the context of the whole LO-doc-JabRef interaction. Try to investigate what `OOVoid` and `OOResult` are used in error handling - and compare them to simply throwing exceptions. Check if these can be re-used at other places where there is interaction with the document. For example, in methods where bibliography is generated.
> 
> Thanks for your response! From my understanding, you are wanting for throwing exceptions to be replaced (where it makes sense and doesn't screw up the workflow) with OOVoid and OOResult error handling methods so that there is less deviation in error handling methods throughout the Open Office components?

Yes, and one could make more of them in case of repeated cases which are not covered by these.

I would also invite you to look if other "good first issues" are free in our project, as they would be easier to work on assuming you all are beginners.

**Comment by mfigueireddo** [comment]
Hi.

I submitted a PR trying to solve the "Separation of concerns: classes such as OOBibBase handle both business logic and UI interactions. We can consider extracting UI-related code into a separate class or layer."

You can see in the PR's descriptions I didn't check "I manually tested my changes in running JabRef (always required)". It's because I didn't fully understand what was supposed to happen when I try to connect with OpenOffice. I create a text file but it keeps showing me "Autodetecting paths" indefinitely. I guess it's a stupid mistake but I wasn't able to figure out yet. I will keep looking for it in the documentation but if anyone can help me at this... 😵‍💫

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/873f3ad2-8a23-43ac-9996-ba6c4a40d75e" />

Also, I didn't check "I checked the [user documentation](https://docs.jabref.org/) for up to dateness and submitted a pull request to our [user documentation repository](https://github.com/JabRef/user-documentation/tree/main/en)" because I have yet to understand about the pull request to your user documentation repository.

Other than that, if there's any suggestion or comment about the modifications I proposed I'd like very much to know.

**Comment by subhramit** [comment]
> You can see in the PR's descriptions I didn't check \"I manually tested my changes in running JabRef (always required)\". It's because I didn't fully understand what was supposed to happen when I try to connect with OpenOffice. I create a text file but it keeps showing me \"Autodetecting paths\" indefinitely. I guess it's a stupid mistake but I wasn't able to figure out yet. I will keep looking for it in the documentation but if anyone can help me at this... 😵‍💫

Hey, your first step should be exploring the software and the concerned feature, and not solving issues in the dark. 
There is documentation available for all this - try googling.
Secondly, please keep your texts self-contained - I'm not sure which PR you're talking about.

Edit - found it. Please use Refs. Instead of Closes, as the PR will not fix the entire issue.

**Comment by subhramit** [comment]
Please also go through our contributing guidelines. We have a certain workflow of issue assignment before PRs.

**Comment by mfigueireddo** [comment]
Sorry for that. Got a little anxious and jumped some steps.

I will spend some time learning more about the project and the workflow.

Thanks for the response.
